docs/src/proposals/optimistic-transaction-propagation-signal.md

```diff
@@ -0,0 +1,109 @@
+
+This list is then is randomly shuffled by stake weight.
+
+Shreds are then retransmitted to up to FANOUT neighbors and up to FANOUT
+children.
+
+## Deterministic retransmission tree
+
+`weighted_shuffle` will use a deterministic seed when
+`enable_deterministic_seed` has been enabled based on the triple (shred slot,
+shred index, leader pubkey):
+
+```
+if enable_deterministic_seed(self.slot(), root_bank) {
+    hashv(&[
+        &self.slot().to_le_bytes(),
+        &self.index().to_le_bytes(),
+        &leader_pubkey.to_bytes(),
+    ])
+```
+
+First, only epoch staked nodes will be considered regardless of presence of
+contact info (and possibly including the validator node itself).
+
+A deterministic ordering of the epoch staked nodes will be created based on the
+derministic shred seed using weighted_shuffle.
+
+Let `neighbor_set` be selected from up to FANOUT neighbors of the current node.
+Let `child_set` be selected from up to FANOUT children of the current node.
+
+Filter `neighbor_set` by contact info.
+Filter `child_set` by contact info.
+
+Let `epoch_set` be the union of `neighbor_set` and `child_set`.
+
+Let `remaining_set` be all other nodes with contact info not contained in
+`epoch_set`.
+
+If `epoch_set.len < 2*FANOUT` then we may randomly select up to
+`2*FANOUT - epoch_set.len` nodes to to retransmit to from `remaining_set`.
+
+## Receiving retransmitted shred
+
+If the current validator node is not in the set of epoch staked nodes for the
+shred epoch then no early retransmission information can be obtained.
+
+Compute the deterministic shred seed.
+
+Run the deterministic epoch_stakes shuffle.
+
+Find position of self in the neighbor or child sets.
+
+Calculate the sum of the stakes of all nodes in the current and prior
+distribution levels.
+
+### Stake summation considerations:
+
+- Stake sum could include stakes of nodes which had been skipped in prior
+distribution levels because of lack of contact info.
+- Current node was part of original epoch staked shuffle from retransmitter
+but was filtered out because of missing contact info. Current node subsequently
+receives retransmisison of shred and assumes that the retransmit was a result
+of the deterministic tree calculation and not from subsequent random selection.
+This should be benign because the current node will underestimate prior stake
+weight in the retransmission tree.
+
+### General considerations:
+
+attack by leader (level 0):
+- transmits shred for distribution through the tree as normal
+- additionally transmits shred (or fake shred) directly to node(s) at level >=2
+leading the node(s) to believe a greater percentage of the tree retransmission
+tree had been processed
+
+attack by node at level n:
+- retransmits shred to node(s) at level >=n+2 leading the node(s) to believe a
+greater percentage of the tree retransmission tree had been processed
+
+### Questions
+
+- Should receiving nodes attempt to verify that the origin of the shred was
```

Comment:
Ah, this is only necessary once a a validator needs to infer from receiving the shred some confidence in the path it went through? Without that need, it doesn't seem like a validator in turbine really cares about where it got the shred from, they just need to decide from the shred seed whether to retransmit the shred.

Could you add a section describing what sort of guarantees we're trying to infer from receiving a shred? I.e. if we receive X shreds via turbine, what can we deduce about the % confidence that the rest of the network will also get this block, and what is the justification (some quick maths/probability calculation) about why this is the case?

Reply:
Yeah, checking the packet source address would be more of a sanity check as it wouldn't prevent spoofing. It's not required for expected operation of the retransmission.

Any signal that would be generated would be advisory. This would indicate the following: if the network is following the above retransmission algorithm then when the current shred was received by this node it should also have been _attempted_ to have been sent to nodes covering `X` stake.

I think a stronger statement would require additional inter-node communication (like some random sampling of nodes that are also expected to have received the shred for confirmation). @aeyakovenko what level of confidence did you have in mind?

Reply:
we can add signatures on hops sometime later.  I think initially I am fine with 
1. assuming UDP spoofing isn't an issue
2. adding private p2p ports between any two nodes on the network, which would make it hard to do 1
3. aggregated signatures on hops with BLS (but that will take a while to ship)

Reply:
Something else we could consider would be to perform the stake distribution calculation on the client side. The validator transmission is made deterministic and the validator simply exposes whether it has received a shred. The client can then perform the stake distribution calculation using validator responses from one or more nodes depending on the client confidence requirements. This would let the client select how much sampling is required for its use case and would offload calculation overhead to the client.